src/models/SAFT/PCSAFT/variants/PPCSAFT.jl

```diff
@@ -73,41 +80,191 @@ function a_res(model ::PPCSAFTModel, V, T, z)
+    iszero(a₂) && return zero(a₂)
+    a₃ = @f(a_3_dq,_data)
     return a₂^2/(a₂-a₃)
 end
 
-function a_2(model ::PPCSAFTModel, V, T, z, _data=@f(data))
+# Deals only with DD & QQ
+function a_2(model ::PPCSAFTModel, V, T, z, type, _data=@f(data))
+    dp_comps, qp_comps = @f(polar_comps)
+    P̄² = []
+    p_comps = []
+    p = 0
+    coeff = 0.
+    if type == "DD"
+        if isempty(dp_comps) return 0. end
+        P̄² = model.params.dipole2.values
+        p_comps = dp_comps
+        p = 3
+        coeff = 1.
+    end
+    if type == "QQ"
+        if isempty(qp_comps) return 0. end
+        P̄² = model.params.quadrupole2.values
+        p_comps = qp_comps
+        p = 7
+        coeff = 9/16
+    end
     _,_,_,_,η,_ = _data
     ∑z = sum(z)
     ρ = N_A*∑z/V
     _a_2 = zero(T+V+first(z))
-    nc = length(model)
+    # nc = length(model)
     m = model.params.segment.values
     ϵ = model.params.epsilon.values
     σ = model.params.sigma.values
-    μ̄² = model.params.dipole2.values
-    @inbounds for i ∈ 1:nc
-        _J2_ii = @f(J2,i,i,η,m,ϵ)
+    # μ̄² = model.params.dipole2.values
+    # Q̄² = model.params.quadrupole2.values
+    @inbounds for (idx, i) ∈ enumerate(p_comps)
+        _J2_ii = @f(J2,type,i,i,η,m,ϵ)
         zᵢ = z[i]
-        μ̄²ᵢ = μ̄²[i]
-        _a_2 +=zᵢ^2*μ̄²ᵢ^2/σ[i,i]^3*_J2_ii
-        for j ∈ (i+1):nc
-            _J2_ij = @f(J2,i,j,η,m,ϵ)
-            _a_2 += 2*zᵢ*z[j]*μ̄²ᵢ*μ̄²[j]/σ[i,j]^3*_J2_ij
+        P̄²ᵢ = P̄²[i]
+        _a_2 +=zᵢ^2*P̄²ᵢ^2/σ[i,i]^p*_J2_ii
+        for j ∈ p_comps[idx+1:end]
+            _J2_ij = @f(J2,type,i,j,η,m,ϵ)
+            _a_2 += 2*zᵢ*z[j]*P̄²ᵢ*P̄²[j]/σ[i,j]^p*_J2_ij
         end
     end
-    _a_2 *= -π*ρ/(T*T)/(∑z*∑z)
+    # @inbounds for i ∈ 1:nc
+    #     _J2_ii = @f(J2,type,i,i,η,m,ϵ)
+    #     zᵢ = z[i]
+    #     μ̄²ᵢ = μ̄²[i]
+    #     _a_2 +=zᵢ^2*μ̄²ᵢ^2/σ[i,i]^3*_J2_ii
+    #     for j ∈ (i+1):nc
+    #         _J2_ij = @f(J2,type,i,j,η,m,ϵ)
+    #         _a_2 += 2*zᵢ*z[j]*μ̄²ᵢ*μ̄²[j]/σ[i,j]^3*_J2_ij
+    #     end
+    # end
+    _a_2 *= -π*coeff*ρ/(T*T)/(∑z*∑z)
     return _a_2
 end
 
-function a_3(model ::PPCSAFTModel, V, T, z, _data=@f(data))
+function a_2_dq(model ::PPCSAFTModel, V, T, z, _data=@f(data))
+    dp_comps, qp_comps = @f(polar_comps)
+    Q̄² = model.params.quadrupole2.values
+    μ̄² = model.params.dipole2.values
+    _,_,_,_,η,_ = _data
+    ∑z = sum(z)
+    ρ = N_A*∑z/V
+    _a_2 = zero(T+V+first(z))
+    m = model.params.segment.values
+    ϵ = model.params.epsilon.values
+    σ = model.params.sigma.values
+    @inbounds for i in dp_comps
+        for j ∈ qp_comps
+            _J2_ij = @f(J2,type,i,j,η,m,ϵ)
+            _a_2 += z[i]*z[j]*μ̄²[i]*Q̄²[j]/σ[i,j]^5*_J2_ij
+        end
+    end
+    _a_2 *= -π*9/4*ρ/(T*T)/(∑z*∑z)
+    return _a_2
+end
+
+function a_3(model ::PPCSAFTModel, V, T, z, type, _data=@f(data))
+    dp_comps, qp_comps = @f(polar_comps)
+    P̄² = []
+    p_comps = []
+    p = 0
+    coeff = 0.
+    if type == "DD"

```

Comment:
try using symbols (`:DD`) instead of strings `"DD"`. symbols don't allocate (at the cost of a slightly higher compilation time)